### PR TITLE
Use haskell-lsp’s builtin VFS in "damlc ide"

### PR DESCRIPTION
**Pull request by cocreature**
haskell-lsp has a builtin VFS that it updates automatically on the
corresponding requests. This PR removes our own VFS implementation and
uses that builtin VFS in "damlc ide". To allow the use of functions
like setBufferModified (we use that heavily in daml-ghc-shake-test-ci)
without having to spawn an LSP server, we also add a fallback where we
spin up our own LSP implementation.

Sorry for the somewhat large diff. I could probably split some of this out into a refactoring PR but I think the refactoring parts on their own are hard to make sense off. Most of the changes are fairly mechanical so hopefully it should be reasonably easy to review.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs [hunk 1]
```diff
@@ -6,44 +6,74 @@
...
-
--- Store the DirtyFiles globally, so we can get at it through setBufferModified
-newtype GlobalDirtyFiles = GlobalDirtyFiles (Var DirtyFiles)
-instance IsIdeGlobal GlobalDirtyFiles
+import Language.Haskell.LSP.Core
+import Language.Haskell.LSP.VFS
+
+-- | haskell-lsp manages the VFS internally and automatically so we cannot use
+-- the builtin VFS without spawning up an LSP server. To be able to test things
+-- like `setBufferModified` we abstract over the VFS implementation.
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs]: I'm not averse to spinning up an LSP server to test, then our tests are more like real. We'd then just mock VS Code. Although happy to go this way for now.

File: compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs [hunk 2]
```diff
@@ -6,44 +6,74 @@
...
+data VFSHandle = VFSHandle
+    { getVirtualFile :: Uri -> IO (Maybe VirtualFile)
+    , setVirtualFileContents :: Uri -> T.Text -> Int -> IO ()
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs]: Is the `Int` required? I guess that's the version? Should it not just always increment?

File: compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs [hunk 3]
```diff
@@ -6,44 +6,74 @@
...
+    makeVFSHandle,
+    makeLSPVFSHandle,
     ) where
 
 
 
 import           StringBuffer
 import Development.IDE.Orphans()
 
+import Control.Concurrent.STM
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs]: You're using STM like they are MVars. I'd prefer direct use of something like a `Var`.

File: compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs (hunk 2 above)
Comment (neil-da) [reply compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs]: In fact I'd just suggest making the `Var` be of type `(Int, Map Uri VirtualFile)` and have a global version.

File: compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs [hunk 4]
```diff
@@ -113,29 +143,21 @@ getFileExists =
...
-setBufferModified :: IdeState -> FilePath -> (Maybe T.Text, UTCTime) -> IO ()
-setBufferModified state absFile (mcontents, !time) = do
-    GlobalDirtyFiles envDirtyFiles <- getIdeGlobalState state
-    -- update vars synchronously
-    modifyVar_ envDirtyFiles $ evaluate . case mcontents of
-        Nothing -> Map.delete absFile
-        Just contents -> Map.insert absFile $ strictPair time $ textToStringBuffer contents
-
-    -- run shake to update results regarding the files of interest
+setBufferModified :: IdeState -> FilePath -> Maybe (T.Text, Int) -> IO ()
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs]: We should drop the `Int` argument.

File: daml-foundations/daml-ghc/daml-compiler/src/DA/Service/Daml/Compiler/Impl/Handle.hs [hunk 5]
```diff
@@ -151,7 +152,12 @@ instance Shake.IsIdeGlobal GlobalPkgMap
 newIdeState :: Options
             -> Maybe (Event -> IO ())
             -> Logger.Handle IO
-            -> Managed IdeState
+            -- This type is rather awkward. The reason for this is that
+            -- we can only fully initialize the IdeState once we get the
+            -- LspFuncs in the initialize callback.
+            -- We should move away from Managed (we only use it for the scenario service)
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-compiler/src/DA/Service/Daml/Compiler/Impl/Handle.hs]: Agreed!

File: daml-foundations/daml-ghc/daml-compiler/src/DA/Service/Daml/Compiler/Impl/Handle.hs [hunk 6]
```diff
@@ -258,12 +279,11 @@ compileFile service fp = do
 onFileModified
     :: IdeState
     -> FilePath
-    -> Maybe T.Text
+    -> Maybe (T.Text, Int)
     -> IO ()
-onFileModified service fp mcontents = do
+onFileModified service fp mbContents = do
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-compiler/src/DA/Service/Daml/Compiler/Impl/Handle.hs]: Would be nice to kill this Int.

File: daml-foundations/daml-ghc/ide/test/Development/IDE/State/API/Testing.hs [hunk 7]
```diff
@@ -94,6 +94,7 @@ data ShakeTestEnv = ShakeTestEnv
     { steTestDirPath :: FilePath -- canonical absolute path of temporary test directory
     , steService :: API.IdeState
     , steVirtualResources :: TVar (Map VirtualResource T.Text)
+    , steFileVersions :: TVar (Map FilePath Int)
```
Comment (neil-da) [inline daml-foundations/daml-ghc/ide/test/Development/IDE/State/API/Testing.hs]: Yuk. If we get rid of the Int stuff we get rid of this too?

File: compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs (hunk 1 above)
Comment (cocreature) [reply compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs]: Yeah, I am certainly planning to do LSP based tests (probably using `lsp-test`). That would also allow us to finally kill those dreaded language server tests. I just didn’t want to also cram this in this PR as well and keeping `daml-ghc-shake-test-ci` intact for now seems useful.

File: compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs (hunk 2 above)
Comment (cocreature) [reply compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs]: The reason why I kept it this way is that it makes it look a bit more like a reasonable implementation that we could theoretically also use in `damlc ide` if we ever switch to something that does not manage the VFS automagically like `haskell-lsp`. But I guess `haskell-lsp` is unlikely to change in that regard unless so changing it to `Int` seems reasonable.

File: compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs (hunk 3 above)
Comment (cocreature) [reply compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs]: Fair enough, I generally default to STM but I don’t feel strongly about this so happy to change it.

**Comment by cocreature** [comment]
I’ve removed the `Int` arguments and everything that required it and switched to `Var` instead of `TVar`. I’ll kill `managed` in a separate PR.

**Pull request merged**